scripts/variantstore/wdl/extract/vds_validation.py

```diff
@@ -44,13 +44,22 @@ def check_densify_small_region(vds):
 
 
 def main(vds):
-    check_samples_match(vds)
-    check_ref_blocks(vds)
-    check_densify_small_region(vds)
+    CHROM_MAP = {'chr1': '1', 'chr2': '2', 'chr3': '3', 'chr4': '4', 'chr5': '6', 'chr6': '6', 'chr7': '7', 'chr8': '8', 'chr9': '9', 'chr10': '10', 'chr11': '11', 'chr12': '12', 'chr13': '13', 'chr14': '14', 'chr15': '15', 'chr16': '16', 'chr17': '17', 'chr18': '18', 'chr19': '19', 'chr20': '20', 'chr21': '21', 'chr22': '22', 'chrX': '23', 'chrY': '24', 'chrM': '25'}
+    for key in CHROM_MAP:
+        chromosome_to_validate = key
+        filtered_vd = vds.variant_data.filter_rows(vds.variant_data.locus.contig == chromosome_to_validate)
+        filtered_rd = vds.reference_data.filter_rows(vds.reference_data.locus.contig == chromosome_to_validate)
+        filtered_vds = hl.vds.VariantDataset(filtered_rd, filtered_vd)
+        # check_samples_match(filtered_vds) # (already succeeded for Echo)
+        print(chromosome_to_validate)
+        check_ref_blocks(filtered_vds)
+
+    # check_densify_small_region(vds) # (already succeeded for Echo)
     vds.validate(); print('Hail VDS validation successful')
```

Comment:
do we think that the vds.validate() needs to be split out by chr too?

Reply:
I'd say so, it's doing very similar things that you are (and some of the same things...)

Reply:
Here is the implementation of `vds.validate`, modified to do all the checks that you do here. You could use this to save a pass over the data and some redundant work.
```python
rd = self.reference_data  
vd = self.variant_data  
  
def error(msg):  
    raise ValueError(f'VDS.validate: {msg}')  
  
rd_row_key = rd.row_key.dtype  
if (  
    not isinstance(rd_row_key, hl.tstruct)  
    or len(rd_row_key) != 1  
    or not rd_row_key.fields[0] == 'locus'  
    or not isinstance(rd_row_key.types[0], hl.tlocus)  
):  
    error(f"expect reference data to have a single row key 'locus' of type locus, found {rd_row_key}")  
  
vd_row_key = vd.row_key.dtype  
if (  
    not isinstance(vd_row_key, hl.tstruct)  
    or len(vd_row_key) != 2  
    or not vd_row_key.fields == ('locus', 'alleles')  
    or not isinstance(vd_row_key.types[0], hl.tlocus)  
    or vd_row_key.types[1] != hl.tarray(hl.tstr)  
):  
    error(  
        f"expect variant data to have a row key {{'locus': locus<rg>, alleles: array<str>}}, found {vd_row_key}"  
    )  
  
rd_col_key = rd.col_key.dtype  
if not isinstance(rd_col_key, hl.tstruct) or len(rd_row_key) != 1 or rd_col_key.types[0] != hl.tstr:  
    error(f"expect reference data to have a single col key of type string, found {rd_col_key}")  
  
vd_col_key = vd.col_key.dtype  
if not isinstance(vd_col_key, hl.tstruct) or len(vd_col_key) != 1 or vd_col_key.types[0] != hl.tstr:  
    error(f"expect variant data to have a single col key of type string, found {vd_col_key}")  
  
if 'END' not in rd.entry or rd.END.dtype != hl.tint32:  
    error("expect field 'END' in entry of reference data with type int32")  
  
if check_data:  
    # check cols
    # PS: you can skip this as you've already verified samples
#    ref_cols = rd.col_key.collect()  
#    var_cols = vd.col_key.collect()  
#    if len(ref_cols) != len(var_cols):  
#        error(  
#            f"mismatch in number of columns: reference data has {ref_cols} columns, variant data has {var_cols} columns"        )  
#  
#    if ref_cols != var_cols:  
#        first_mismatch = 0  
#        while ref_cols[first_mismatch] == var_cols[first_mismatch]:  
#            first_mismatch += 1  
#        error(  
#            f"mismatch in columns keys: ref={ref_cols[first_mismatch]}, var={var_cols[first_mismatch]} at position {first_mismatch}"  
#        )  
  
    # check locus distinctness  
    n_rd_rows = rd.count_rows()  
    n_distinct = rd.distinct_by_row().count_rows()  
  
    if n_distinct != n_rd_rows:  
        error(f'reference data loci are not distinct: found {n_rd_rows} rows, but {n_distinct} distinct loci')  
  
    # check END field  
    (missing_end, end_before_position, block_too_long, zero_gq) = rd.aggregate_entries((  
        hl.agg.filter(hl.is_missing(rd.END), hl.agg.take((rd.row_key, rd.col_key), 5)),  
        hl.agg.filter(rd.END < rd.locus.position, hl.agg.take((rd.row_key, rd.col_key), 5)),  
        # PS: extra checks added here
        hl.agg.filter(rd.END - rd.locus_start >= rd.ref_block_max_length, hl.agg.take((rd.row_key, rd.col_key), 5)),
        hl.agg.filter(rd.GQ <= 0, hl.agg.take((rd.row_key, rd.col_key), 5))
    ))  
  
    if missing_end:  
        error(  
            'found records in reference data with missing END field\n  '
            + '\n  '.join(str(x) for x in missing_end)  
        )  
    if end_before_position:  
        error(  
            'found records in reference data with END before locus position\n  '
            + '\n  '.join(str(x) for x in end_before_position)  
        )
    if block_too_long:  
        error(  
            'found records in reference data with block length greater than max length\n  '
            + '\n  '.join(str(x) for x in block_too_long)  
        )
    if zero_gq:  
        error(  
            'found records in reference data with GQ <= 0\n  '            
            + '\n  '.join(str(x) for x in zero_gq)  
        )
```

Reply:
I think we'll be modifying this so that next time, you can just call `vds.validate` and pass in the `rd.GQ > 0` check.